technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,110 @@
+# Medusa Operator - Restore redesign document
+
+## Background
+
+Medusa operator doesn’t yet support remote restores: restoring a backup to another cluster.
+As we implement this feature, we would like to have it support multiple use cases that operators could expect to have working:
+
+- Restoring a medusa-operator based backup to another Cassandra install in the same k8s cluster
+![](img/medusa_k8s-new-restore-design-1.jpg)
+
+- Restoring a medusa-operator based backup to another Cassandra install in a different k8s cluster
+![](img/medusa_k8s-new-restore-design-2.jpg)
+
+- Restoring a Medusa backup (non k8s) to a k8s Cassandra install
+![](img/medusa_k8s-new-restore-design-3.jpg)
+
+
+
+
+## Sync CassandraBackup resources with storage
+The target solution will need to be able to sync the local CassandraBackup resources with the configured storage backend.
+The sync should allow:
+
+- Creating local missing resources from existing backups
+- Deleting local resources for missing backups
+- Deleting remote backups if a local resource was marked as deleted
+
+A single control loop should be able to handle all 3 operations.
+The sync should be required explicitly, meaning that a new CRD should be created to request synchronization to the medusa-operator, and that CRD should contain a prefix so that a cluster can list backups stored for another cluster in a multi-tenant storage bucket.
+
+The CassandraBackup resource should be updated, adding a delete field which would allow marking backups for deletion.
+
+The sync operation would:
+
+- List the backups in the storage bucket, through a gRPC call
+- Compare that list to the existing CassandraBackup resources
+- Create a CassandraBackup object for each missing backup
+- Delete local resources that don’t exist in the list of backups
+- Delete remote backups through gRPC for CassandraBackup objects marked for deletion, and upon success delete the resource.
+
+## CassandraBackup CRD changes
+CassandraBackup resources shouldn’t store the CassandraDatacenter spec anymore but rather the topology object generated by Medusa.  
+This object will contain the list of live nodes at the time of the backup along with their token assignments.  
+The CRD should also contain information that should be overridable such as the prefix, which is controlled by the namespace name. 
+
+## Medusa changes
+The tokenmap generated by Medusa will need to be augmented to provide the placement of nodes in the cluster: datacenter and rack.
+Currently Medusa is not dc/rack aware and it can probably be fairly surprising to users. It was built to backup a single DC per cluster and didn’t account for racks either.  
+Restoring from a DC/rack aware cluster to another one imposes some constraints on the source/target nodes mapping, which are currently impossible to deal with safely in a remote restore.
```

Comment:
What are those constraints?

Reply:
When using NetworkTopologyStrategy, the placement of replicas is dependent on the DC and rack layout.
In this context, the replica for a token range will be placed on the node that owns the next token range in the next rack.
We can't restore any node to any other node as we may end up putting several replicas in the same rack, and the nodes we'd restore to would possibly end up not containing data they're supposed be replica for.
We can be rigid about the mapping and expect the same dc and rack names exclusively, or do a 1:1 mapping between DCs/Racks despite having different names (but expect the same number of both).

Reply:
Thanks for the detailed explanation!

Reply:
Thanks, I've added the explanation to the doc itself.